### PR TITLE
Introduce new PYTEST_CURRENT_TEST environment variable

### PR DESCRIPTION
**Pull request by nicoddemus**
Introduce new `PYTEST_CURRENT_TEST` environment variable (#2583).

While looking at the docs I decided to improve the section about selecting tests because `PYTEST_CURRENT_TEST` is all about `nodeid`s and I didn't find a good explanation about that.

Also improve the docs on how `PYTEST_ADDOPTS` and `addopts` interact to create the final command-line.

### CODE REVIEW

File: doc/en/example/simple.rst [hunk 1]
```diff
@@ -761,6 +761,36 @@ and run it::
...
+
+.. versionadded:: 3.2
+
+Sometimes a test session might get stuck and there might be no easy way to figure out
+which test got stuck, for example if pytest was run in quiet mode (``-q``) or you don't have access to the console
+output. This is particularly a problem if the problem helps only sporadically, the famous "flaky" kind of tests.
+
+``pytest`` sets a ``PYTEST_CURRENT_TEST`` environment variable when running tests, which can be inspected
+by process monitoring utilities or libraries like `psutil <https://pypi.python.org/pypi/psutil>`_ to discover which
+test got stuck if necessary.
```
Comment (The-Compiler) [inline doc/en/example/simple.rst]: I assume that'd be something like `psutil.Process(pid).environ()['PYTEST_CURRENT_TEST']` then? Maybe add that as an example?

File: doc/en/example/simple.rst [hunk 2]
```diff
@@ -761,6 +761,36 @@ and run it::
...
+
+#. ``foomodule.py::test_foo,setup``
+#. ``foomodule.py::test_foo,call``
+#. ``foomodule.py::test_foo,teardown``
+
+In that order.
+
+During each ``setup``, ``call`` and ``teardown`` test stage, pytest will set ``PYTEST_CURRENT_TEST`` to the items'
+:ref:`nodeid <nodeids>` and stage
+``<nodeid>,<stage>``. As an example, a test with node id
```
Comment (The-Compiler) [inline doc/en/example/simple.rst]: this seems outdated

Comment (nicoddemus) [inline doc/en/example/simple.rst]: Oops! Indeed, thanks!

File: doc/en/example/simple.rst [hunk 3]
```diff
@@ -761,6 +761,41 @@ and run it::
...
+
+Sometimes a test session might get stuck and there might be no easy way to figure out
+which test got stuck, for example if pytest was run in quiet mode (``-q``) or you don't have access to the console
+output. This is particularly a problem if the problem helps only sporadically, the famous "flaky" kind of tests.
+
+``pytest`` sets a ``PYTEST_CURRENT_TEST`` environment variable when running tests, which can be inspected
+by process monitoring utilities or libraries like `psutil <https://pypi.python.org/pypi/psutil>`_ to discover which
+test got stuck if necessary:
+
+.. code-block:: python
```
Comment (nicoddemus) [inline doc/en/example/simple.rst]: Added an example as suggested by @The-Compiler, thanks

**Comment by RonnyPfannschmidt** [comment]
for extra future-proofing it may be sensible to go as far as putting each packages metadata into a key that equals the package name (that way pytest plugins can use the same mechanism)

**Comment by nicoddemus** [comment]
> if we put more data than the node id then we shouldn't use poor mans parseable data, but go for json straight away

I had that feeling as well that adding that kind of metadata separated by `,` was a little dubious at best, decided to implement it anyway and get feedback, thanks. 

But while I like using JSON for it, I'm not so sure we should put JSON into an environment variable: it feels weird plus OSs might have a size limit to what you put in there (I know Windows does). I think that job may be best left for plugins that put the whole test suite information in a file, like `pytest-tap`, `result-log` and `jxmlunit`. As you said in previous occasions, we could devise a new plugin which spits this information out in a line-based JSON format, which could even add additional hooks so plugins and users can add additional information, like the discussion we had recently in #2570.

So I propose to remove the `stage` from the environment variable and only leave the `nodeid` from this PR. What do you guys think?

**Comment by The-Compiler** [comment]
I guess the goal of this is mostly it being *human* readable, no (for the case you mentioned where a test is hanging)? There are better ways available to integrate programmatically.

For the usecase mentioned ("pytest hangs, what was it doing?", knowing the stage is very useful. I don't think the format matters much, but since the audience is human, might as well go for `test_foo.py::test (setup)` or so.

-1 for introducing JSON, and -1 for dropping it.

**Comment by nicoddemus** [comment]
> I guess the goal of this is mostly it being human readable

That's a good point. Using `"test_foo.py::test (setup)"` indeed makes it clear the intent is for humans to read, not to parse it. We might go further and make it explicit in the docs that this is the intent and it can change in any pytest release, including bug fixes.

What do you think @RonnyPfannschmidt?

**Comment by RonnyPfannschmidt** [comment]
Sounds fine that way

**Comment by nicoddemus** [comment]
Thanks, will update the PR then!

**Comment by nicoddemus** [comment]
Updated.

**Comment by The-Compiler** [comment]
Thanks!

**Pull request merged**